src/trace/cursor/cursor_list.rs

```diff
@@ -104,93 +105,108 @@ where
-        debug_assert!(self.cursors[self.min_val[0]].val_valid(&storage[self.min_val[0]]));
-        self.cursors[self.min_val[0]].val(&storage[self.min_val[0]])
+    fn val(&self) -> &V {
+        debug_assert!(self.key_valid());
+        debug_assert!(self.val_valid());
+        debug_assert!(self.cursors[self.min_val[0]].val_valid());
+        self.cursors[self.min_val[0]].val()
     }
     #[inline]
-    fn map_times<L: FnMut(&T, &R)>(&mut self, storage: &Self::Storage, mut logic: L) {
+    fn map_times<L: FnMut(&T, &R)>(&mut self, mut logic: L) {
         for &index in self.min_val.iter() {
-            self.cursors[index].map_times(&storage[index], |t, d| logic(t, d));
+            self.cursors[index].map_times(|t, d| logic(t, d));
         }
     }
 
     #[inline]
-    fn weight(&mut self, storage: &Self::Storage) -> R
+    fn weight(&mut self) -> R
     where
         T: PartialEq<()>,
     {
-        debug_assert!(self.key_valid(storage));
-        debug_assert!(self.val_valid(storage));
-        debug_assert!(self.cursors[self.min_val[0]].val_valid(&storage[self.min_val[0]]));
+        debug_assert!(self.key_valid());
+        debug_assert!(self.val_valid());
+        debug_assert!(self.cursors[self.min_val[0]].val_valid());
         let mut res: R = HasZero::zero();
-        self.map_times(storage, |_, w| res.add_assign_by_ref(w));
+        self.map_times(|_, w| res.add_assign_by_ref(w));
         res
     }
 
     // key methods
     #[inline]
-    fn step_key(&mut self, storage: &Self::Storage) {
+    fn step_key(&mut self) {
         for &index in self.min_key.iter() {
-            self.cursors[index].step_key(&storage[index]);
+            self.cursors[index].step_key();
         }
-        self.minimize_keys(storage);
+        self.minimize_keys();
     }
     #[inline]
-    fn seek_key(&mut self, storage: &Self::Storage, key: &K) {
-        for (index, cursor) in self.cursors.iter_mut().enumerate() {
-            cursor.seek_key(&storage[index], key);
+    fn seek_key(&mut self, key: &K) {
+        for cursor in self.cursors.iter_mut() {
+            cursor.seek_key(key);
         }
-        self.minimize_keys(storage);
+        self.minimize_keys();
     }
 
     // value methods
     #[inline]
-    fn step_val(&mut self, storage: &Self::Storage) {
+    fn step_val(&mut self) {
         for &index in self.min_val.iter() {
-            self.cursors[index].step_val(&storage[index]);
+            self.cursors[index].step_val();
         }
-        self.minimize_vals(storage);
+        self.minimize_vals();
     }
     #[inline]
-    fn seek_val(&mut self, storage: &Self::Storage, val: &V) {
+    fn seek_val(&mut self, val: &V) {
         for &index in self.min_key.iter() {
-            self.cursors[index].seek_val(&storage[index], val);
+            self.cursors[index].seek_val(val);
+        }
+        self.minimize_vals();
+    }
+
+    fn values<'a>(&mut self, vals: &mut Vec<(&'a V, R)>)
```

Comment:
why is it better to fill an existing vector than return a new one?

Reply:
is it documented that this does not return 0 values?

Reply:
Performance-wise, it's better because it reduces memory allocations.  In this case, we also need this functionality so we can assemble values from multiple cursors in one vector.